File: saat/selection.py
"""Milestone 20: watch-selection logic for the today picker and the week
planner. Pure — no Qt, no I/O, no event loop. See SPEC.md §5.5.

Both modes draw from wear.owned_watches() only: SPEC.md §5.12 excludes
Wishlist, Incoming, Sold and Gifted watches from anything wear-related, the
same choke point Stats mode and the calendar already read through.
"""

import random
from datetime import date, timedelta

from saat.storage import WatchRecord
from saat.ui.wear_stats import last_worn
from saat.wear import owned_watches

MODE_RANDOM = "random"
MODE_WEIGHTED = "weighted"

# A never-worn watch's weight is set this far above whatever the most-
# neglected *real* value in the batch is, rather than a fixed always-huge
# constant. A collection-relative bonus keeps "never worn is favoured" true
# without making a recently-worn watch's chance so small it stops feeling
# like "never zero" in practice — SPEC.md milestone 20 step 3 wants a
# gentle curve specifically so the result never feels rigged.
_NEVER_WORN_BONUS = 30.0


def _weights_from_last_worn(last_worn_by_slug: dict[str, date | None], reference: date) -> dict[str, float]:
    """Shared weighting core for compute_weights() and the weighted week
    roll (which recomputes this against a different reference day per
    iteration as picks accumulate — see pick_week). Weight is linear in
    days-since `reference` — a gentle curve, not a hard cutoff — floored at
    0 days-since rather than going negative, which a pre-planned future
    wear date (SPEC.md §5.5: "Every day is editable, past or future") could
    otherwise produce. Every result is >= 1.0: never zero."""
    since_by_slug = {
        slug: (None if when is None else max(0.0, float((reference - when).days)))
        for slug, when in last_worn_by_slug.items()
    }
    real_values = [v for v in since_by_slug.values() if v is not None]
    never_worn_weight = max(real_values, default=0.0) + _NEVER_WORN_BONUS
    return {slug: (never_worn_weight if since is None else since + 1.0) for slug, since in since_by_slug.items()}


def compute_weights(watches: list[WatchRecord], today: date | None = None) -> dict[str, float]:
    """slug -> weight for weighted mode. Never-worn always gets the maximum
    (see _weights_from_last_worn); the least-recently-worn watch gets the
    least, but never zero."""
    today = today if today is not None else date.today()
    return _weights_from_last_worn({r.slug: last_worn(r.watch) for r in watches}, today)


def pick_random(records: list[WatchRecord], rand: random.Random | None = None) -> WatchRecord:
    """Uniform over owned watches — a true dN. Raises ValueError with no
    owned watches; callers check emptiness before presenting the picker
    (SPEC.md milestone 20 step 11) rather than relying on this exception."""
    watches = owned_watches(records)
    if not watches:
        raise ValueError("no owned watches to pick from")
    rand = rand if rand is not None else random.Random()
    return rand.choice(watches)


def pick_weighted(
    records: list[WatchRecord], rand: random.Random | None = None, today: date | None = None
) -> WatchRecord:
    """Favours watches worn least recently — see compute_weights() for the
    curve. Raises ValueError with no owned watches; see pick_random()."""
    watches = owned_watches(records)
    if not watches:
        raise ValueError("no owned watches to pick from")
    rand = rand if rand is not None else random.Random()
    weights = compute_weights(watches, today)
    return rand.choices(watches, weights=[weights[r.slug] for r in watches], k=1)[0]


def pick_one(
    records: list[WatchRecord], mode: str, rand: random.Random | None = None, today: date | None = None
) -> WatchRecord:
    """Dispatches to pick_random()/pick_weighted() by mode string."""
    if mode == MODE_WEIGHTED:
        return pick_weighted(records, rand, today)
    if mode == MODE_RANDOM:
        return pick_random(records, rand)
    raise ValueError(f"unknown picker mode: {mode!r}")


def week_dates(week_start: date) -> list[date]:
    """The seven dates Monday..Sunday of week_start's week. week_start need
    not itself be a Monday — normalised here the same way month_grid.py's
    month_grid_days() normalises to whole weeks (SPEC.md §5.5: weeks start
    Monday)."""
    monday = week_start - timedelta(days=week_start.weekday())
    return [monday + timedelta(days=i) for i in range(7)]


def pick_week(
    records: list[WatchRecord],
    week_start: date,
    mode: str,
    rand: random.Random | None = None,
    today: date | None = None,
) -> dict[date, WatchRecord]:
    """One pick per day, Monday..Sunday of week_start's week. SPEC.md
    milestone 20 step 5: repeats are allowed by default (a <=7-watch
    collection cannot fill a week otherwise), but with more than seven owned
    watches, picks prefer distinct watches — a hard guarantee in random mode
    (sampling without replacement), a strong-but-never-absolute preference
    in weighted mode, since hard-excluding an already-picked watch on a
    later day would zero out its chance and break the same never-zero-
    weight invariant compute_weights() promises everywhere else.

    Returns picks for all seven days unconditionally — this function has no
    opinion on which days are empty, already logged, or in the past. That
    filtering belongs to the caller (the week planner in calendar_view.py),
    which decides which of these seven proposals to actually offer or
    write. Raises ValueError with no owned watches; see pick_random()."""
    watches = owned_watches(records)
    if not watches:
        raise ValueError("no owned watches to pick from")
    rand = rand if rand is not None else random.Random()
    days = week_dates(week_start)

    if mode == MODE_RANDOM:
        chosen = rand.sample(watches, k=7) if len(watches) >= 7 else [rand.choice(watches) for _ in days]
        return dict(zip(days, chosen))

    if mode == MODE_WEIGHTED:
        # A working copy of "when was this watch last worn", updated as the
        # week fills in via _weights_from_last_worn (same core compute_weights
        # uses), so a watch picked for Monday reads as freshly-worn when
        # Tuesday's weights are computed — carrying the bias across the week
        # (step 16) without ever hard-excluding it.
        last_worn_by_slug: dict[str, date | None] = {r.slug: last_worn(r.watch) for r in watches}
        picks: dict[date, WatchRecord] = {}
        for day in days:
            weights = _weights_from_last_worn(last_worn_by_slug, day)
            chosen = rand.choices(watches, weights=[weights[r.slug] for r in watches], k=1)[0]
            picks[day] = chosen
            last_worn_by_slug[chosen.slug] = day
        return picks

    raise ValueError(f"unknown picker mode: {mode!r}")

File: android/app/src/main/kotlin/io/github/sudomegas/saat/storage/Selection.kt
package io.github.sudomegas.saat.storage

import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.random.Random

/**
 * "Pick for me" — SPEC-ANDROID 5.5, ported from the desktop's
 * `saat/selection.py`. Pure — no ViewModel, no Compose, no I/O. Only the
 * single-day picker is ported; the desktop's week planner (`pick_week`) is
 * explicitly out of scope.
 */

/**
 * A never-worn watch's weight sits this far above whatever the most-neglected
 * *real* value in the batch is, rather than a fixed always-huge constant — a
 * collection-relative bonus so "never worn is favoured" holds without making
 * a recently-worn watch's chance so small it stops feeling like "never zero."
 */
private const val NEVER_WORN_BONUS = 30.0

/** Random vs Weighted. [token] is what reaches `config.toml`'s `[picker]` table. */
enum class PickerMode(val token: String) {
    RANDOM("random"),
    WEIGHTED("weighted"),
    ;

    companion object {
        val DEFAULT: PickerMode = RANDOM

        fun fromToken(token: String?): PickerMode =
            entries.firstOrNull { it.token == token } ?: DEFAULT
    }
}

/**
 * Callers check `records.ownedWatches().isEmpty()` before presenting the
 * picker at all (see `CalendarViewModel.openPickForMe`), exactly as the
 * desktop dialog special-cases an empty collection, rather than relying on
 * this being caught.
 */
class NoOwnedWatchesException : IllegalStateException("no owned watches to pick from")

/**
 * slug -> weight for weighted mode. [watches] must already be owned-only
 * (see [ownedWatches]) — this function does no filtering of its own, mirroring
 * the desktop's `compute_weights`/`_weights_from_last_worn` split.
 *
 * Weight is linear in days-since-[today], floored at 0 rather than going
 * negative — a pre-planned future wear date must not do so, since
 * [Watch.daysSinceWorn] itself deliberately does not floor (SPEC-ANDROID 5.5:
 * "every day is editable, past or future"). Every result is > 0: never worn
 * gets the collection-relative [NEVER_WORN_BONUS], everyone else gets
 * `daysSince + 1.0`, so nothing is ever truly excluded.
 */
fun computeWeights(watches: List<WatchRecord>, today: LocalDate): Map<String, Double> {
    val daysSince: Map<String, Double?> = watches.associate { record ->
        val last = record.watch?.lastWorn()
        val since = last?.let { maxOf(0.0, ChronoUnit.DAYS.between(it, today).toDouble()) }
        record.slug to since
    }
    val neverWornWeight = (daysSince.values.filterNotNull().maxOrNull() ?: 0.0) + NEVER_WORN_BONUS
    return daysSince.mapValues { (_, since) -> since?.plus(1.0) ?: neverWornWeight }
}

/**
 * Uniform over owned watches — a true dN. Throws [NoOwnedWatchesException]
 * with no owned watches; see the class doc for why callers check first.
 */
fun pickRandom(records: List<WatchRecord>, random: Random = Random.Default): WatchRecord {
    val owned = records.ownedWatches()
    if (owned.isEmpty()) throw NoOwnedWatchesException()
    return owned[random.nextInt(owned.size)]
}

/**
 * Favours watches worn least recently — see [computeWeights] for the curve.
 * Throws [NoOwnedWatchesException] with no owned watches; see [pickRandom].
 *
 * Cumulative-weight sampling: draw a point in `[0, total)`, walk the running
 * sum, return the watch whose slice the draw landed in. `total >= 1.0` always
 * (every weight is > 0), so there is no divide-by-zero case — the trailing
 * fallback exists only for floating-point drift leaving the final comparison
 * a hair short of `total`.
 */
fun pickWeighted(
    records: List<WatchRecord>,
    today: LocalDate,
    random: Random = Random.Default,
): WatchRecord {
    val owned = records.ownedWatches()
    if (owned.isEmpty()) throw NoOwnedWatchesException()
    val weights = computeWeights(owned, today)
    val total = owned.sumOf { weights.getValue(it.slug) }
    val draw = random.nextDouble() * total
    var cumulative = 0.0
    for (record in owned) {
        cumulative += weights.getValue(record.slug)
        if (draw < cumulative) return record
    }
    return owned.last()
}

/** Dispatches to [pickRandom]/[pickWeighted] by [mode]. */
fun pickOne(
    records: List<WatchRecord>,
    mode: PickerMode,
    today: LocalDate,
    random: Random = Random.Default,
): WatchRecord = when (mode) {
    PickerMode.RANDOM -> pickRandom(records, random)
    PickerMode.WEIGHTED -> pickWeighted(records, today, random)
}
